### ISSUE TITLE
 Null key for a Map not allowed in JSON  org.keycloak.protocol.oidc.TokenManager$AccessTokenResponseBuilder.build(TokenManager.java:1224)

### ISSUE THREAD

**Issue by jsemohub** [keycloak/keycloak#15467]
### Area

authentication

### Describe the bug

Using keycloak-js to authenticate against keycloak
quay.io/keycloak/keycloak:20.0.1

Token request method fails
http://XXX/realms/test/protocol/openid-connect/token -> receives 500 status response from the server

Keycloak server logs captures jackson parsing exception

eption: com.fasterxml.jackson.databind.JsonMappingException: Null key for a Map not allowed in JSON (use a converting NullKeySerializer?) (throug
h reference chain: org.keycloak.representations.AccessToken["[anySetter]"]->java.util.HashMap["null"])
        at org.keycloak.jose.jws.JWSBuilder.jsonContent(JWSBuilder.java:65)
        at org.keycloak.jose.jws.DefaultTokenManager.encode(DefaultTokenManager.java:79)
        at org.keycloak.protocol.oidc.TokenManager$AccessTokenResponseBuilder.build(TokenManager.java:1224)

### Version

quay.io/keycloak/keycloak:20.0.1

### Expected behavior

Token request should succeed.

### Actual behavior

Token request fails with a 500 server response.

Keycloak server logs captures jackson parsing exception

eption: com.fasterxml.jackson.databind.JsonMappingException: Null key for a Map not allowed in JSON (use a converting NullKeySerializer?) (throug
h reference chain: org.keycloak.representations.AccessToken["[anySetter]"]->java.util.HashMap["null"])
        at org.keycloak.jose.jws.JWSBuilder.jsonContent(JWSBuilder.java:65)
        at org.keycloak.jose.jws.DefaultTokenManager.encode(DefaultTokenManager.java:79)
        at org.keycloak.protocol.oidc.TokenManager$AccessTokenResponseBuilder.build(TokenManager.java:1224)

### How to Reproduce?

1. Create a test realm.
2. Create a test client. Disable authentication.
3. Create a new client scope with two mappers: a) group membership, 2) audience with included and custom audiences referencing client name from previous step. 
4. Add new client scope to the client
5. configure keycloak-js client as per https://www.keycloak.org/docs/latest/securing_apps/index.html#_javascript_adapter
const keycloak = new Keycloak({
url: 'http://keycloakserver',
  realm: 'test',
  clientId: 'test'
})
6. Invoke keycloak.login() in javascript and monitor browser's console. 
7. Observe the error describe above 

### Anything else?

2022-11-11 13:28:01,444 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-99) Uncaught server error: java.lang.RuntimeExc
eption: com.fasterxml.jackson.databind.JsonMappingException: Null key for a Map not allowed in JSON (use a converting NullKeySerializer?) (throug
h reference chain: org.keycloak.representations.AccessToken["[anySetter]"]->java.util.HashMap["null"])
        at org.keycloak.jose.jws.JWSBuilder.jsonContent(JWSBuilder.java:65)
        at org.keycloak.jose.jws.DefaultTokenManager.encode(DefaultTokenManager.java:79)
        at org.keycloak.protocol.oidc.TokenManager$AccessTokenResponseBuilder.build(TokenManager.java:1224)
        at org.keycloak.protocol.oidc.endpoints.TokenEndpoint.createTokenResponse(TokenEndpoint.java:472)
        at org.keycloak.protocol.oidc.endpoints.TokenEndpoint.codeToToken(TokenEndpoint.java:440)
        at org.keycloak.protocol.oidc.endpoints.TokenEndpoint.processGrantRequest(TokenEndpoint.java:204)
        at jdk.internal.reflect.GeneratedMethodAccessor414.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
        at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
        at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:192)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:152)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:183)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:141)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:32)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)

**Comment by jsemohub** [comment]
Worth mentioning that I was testing against the docker image as per
https://www.keycloak.org/getting-started/getting-started-docker

**Comment by jsemohub** [comment]
Removing mappers eliminates the issue.

**Comment by lexcao** [comment]
> Create a new client scope with two mappers: a) group membership, 2) audience with included and custom audiences referencing client name from previous step.

Which mapper do you remove that eliminates the issue?

Could you provide more info about the configuration of the mapper?

I am going to reproduce this from my local.

**Comment by mposolda** [comment]
Thanks for the report, but unfortunately due the amount of other reported issues and other priorities, Keycloak team does not have time to properly triage this bug. So preliminary added to Backlog for now.
It will be helpful if:
- You can verify if still applicable in latest Keycloak released version. If not, then it is welcome to close this issue.
- If you figure that this may not be a valid bug (for example just a mistake in configuration etc), it will be also welcome to close this issue
- If still applicable in latest version, it will be welcome to add the comment as well, that this was still reproduced with latest Keycloak version as it is very valuable info. Anyone is welcome to comment with this or add other relevant comments to this issue.

**Comment by onlyati** [comment]
> Thanks for the report, but unfortunately due the amount of other reported issues and other priorities, Keycloak team does not have time to properly triage this bug. So preliminary added to Backlog for now. It will be helpful if:
> 
> * You can verify if still applicable in latest Keycloak released version. If not, then it is welcome to close this issue.
> * If you figure that this may not be a valid bug (for example just a mistake in configuration etc), it will be also welcome to close this issue
> * If still applicable in latest version, it will be welcome to add the comment as well, that this was still reproduced with latest Keycloak version as it is very valuable info. Anyone is welcome to comment with this or add other relevant comments to this issue.

Hi,

I didn't open this issue, but I have met with the same issue and I also have found a workaround. I am using the current container which tagged as "latest" (that is 21.0 I think). I have got the same behavior than it is in the original post.

First, I was creating this scope for a confidential client, it worked. This is settings of my map:
![image](https://user-images.githubusercontent.com/14370881/230942368-bdc95f89-29da-4977-8e5c-116e93b58989.png)

It is working fine without any additional parameter. But then I applied it to a public client and I have got the errors.
```
2023-04-10 15:36:36,203 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-2) Uncaught server error: java.lang.RuntimeException: com.fasterxml.jackson.databind.JsonMappingException: Null key for a Map not allowed in JSON (use a converting NullKeySerializer?) (through reference chain: org.keycloak.representations.AccessToken["[anySetter]"]->java.util.HashMap["null"])
```

Later, I specified the "Token Claim Name":
![image](https://user-images.githubusercontent.com/14370881/230942703-7a516412-619e-410a-8588-212cbb6ab2b1.png)

And it worked like a charm!

In case of confidential client, it seems that if "Token Claim Name" is not specified, then the "Name" is the default. But in case of public client, it throw an null reference exception. I hope my comment helps and you can recreate it.

**Comment by rmartinc** [comment]
Probably this is fixed by the PR: https://github.com/keycloak/keycloak/pull/20480

Previously the list of groups was directly added into the token and if the claim name was null there was an error. Now the list is simply not added. Can you test with last keycloak 23.0.5?

Nevertheless maybe the "Token Claim Name" should be changed to required. Don't know if that is feasible and/or easy. For this point I'm moving to backlog, PRs are welcome!

**Comment by edewit** [comment]
~old-release

**Comment by rmartinc** [comment]
I'm closing the issue as I think it's already fixed by #20480. Please re-open or file a new one if I'm wrong.
